mgmtsystem/azure.py

```diff
@@ -484,6 +593,36 @@ def current_ip_address(self, vm_name, resource_group=None):
         return etree.parse(StringIO(data)).getroot().xpath(
             "./Object/Property[@Name='IpAddress']/text()")
 
+    def list_free_nics(self, nic_template):
+        try:
+            azure_data = self.run_script(
+                """
+                Get-AzureRmNetworkInterface -ResourceGroupName {}|
+                Where-Object {{$_.VirtualMachineText -eq "null" -and $_.Name -like "{}"}}|
+                convertto-xml -as String
+                """.format(self.resource_group, nic_template))
+            data = self.clean_azure_xml(azure_data)
+            nic_list = etree.parse(StringIO(data)).getroot().xpath(
+                "./Object/Property[@Name='Name']/text()")
+            return nic_list
+        except ActionTimedOutError:
+            return False
+
+    def list_free_pip(self, pip_template):
+        try:
+            azure_data = self.run_script(
+                """
+                Get-AzureRmPublicIpAddress -ResourceGroupName {}|
+                Where-Object {{$_.Name -like "{}" -and $_.IpConfigurationText -eq "null"}}|
+                convertto-xml -as String
+                """.format(self.resource_group, pip_template))
+            data = self.clean_azure_xml(azure_data)
+            pip_list = etree.parse(StringIO(data)).getroot().xpath(
+                "./Object/Property[@Name='Name']/text()")
+            return pip_list
+        except ActionTimedOutError:
+            return False
+
```

Comment:
The committee can squash the commits if needed :) But already squashed is always preferable.